src/rgw/s3select.h

```diff
@@ -0,0 +1,852 @@
+
+            alias_name = lexeme_d[(+alpha_p >> *digit_p)] ;
+
+
+            s3_object = str_p("stdin") | object_path ; 
+
+            object_path = "/" >> *( fs_type >> "/") >> fs_type;
+
+            fs_type = lexeme_d[+( alnum_p | str_p(".")  | str_p("_")) ];
+
+            where_clause = str_p("where") >> condition_expression;
+
+            condition_expression = (arithmetic_predicate >> *(log_op[BOOST_BIND_ACTION(push_logical_operator)] >> arithmetic_predicate[BOOST_BIND_ACTION(push_logical_predicate)]));
+
+            arithmetic_predicate = (factor >> *(arith_cmp[BOOST_BIND_ACTION(push_compare_operator)] >> factor[BOOST_BIND_ACTION(push_arithmetic_predicate)]));
+
+            factor = (arithmetic_expression) | ('(' >> condition_expression >> ')') ; 
+
+            arithmetic_expression = (addsub_operand >> *(addsubop_operator[BOOST_BIND_ACTION(push_addsub)] >> addsub_operand[BOOST_BIND_ACTION(push_addsub_binop)] ));
+
+            addsub_operand = (mulldiv_operand >> *(muldiv_operator[BOOST_BIND_ACTION(push_mulop)]  >> mulldiv_operand[BOOST_BIND_ACTION(push_mulldiv_binop)] ));// this non-terminal gives precedense to  mull/div
+
+            mulldiv_operand = arithmetic_argument | ('(' >> (arithmetic_expression) >> ')') ; 
+
+            list_of_function_arguments = (arithmetic_expression)[BOOST_BIND_ACTION(push_function_arg)] >> *(',' >> (arithmetic_expression)[BOOST_BIND_ACTION(push_function_arg)]) ;
+
+            function =( (variable >>  '(' )[BOOST_BIND_ACTION(push_function_name)]   >> list_of_function_arguments >> ')' ) [BOOST_BIND_ACTION(push_function_expr)];
+            
+            arithmetic_argument = (float_number)[BOOST_BIND_ACTION(push_float_number)] |  (number)[BOOST_BIND_ACTION(push_number)] | (column_pos)[BOOST_BIND_ACTION(push_column_pos)] | 
+                                (string)[BOOST_BIND_ACTION(push_string)] |
+                                (function)[BOOST_BIND_ACTION(push_debug_1)]  | (variable)[BOOST_BIND_ACTION(push_variable)] ;//function is pushed by right-term 
+
+                       
+            number = int_p;
+
+            float_number = real_p;
+
+            string = str_p("\"") >> *( anychar_p - str_p("\"") ) >> str_p("\"") ;
+
+            column_pos = ('_'>>+(digit_p) ) | '*' ;
+
+            muldiv_operator = str_p("*") | str_p("/") | str_p("^");// got precedense
+
+            addsubop_operator = str_p("+") | str_p("-");
+
+
+            arith_cmp = str_p(">=") | str_p("<=") | str_p("==") | str_p("<") | str_p(">") | str_p("!=");
+
+            log_op = str_p("and") | str_p("or"); //TODO add NOT (unary)
+
+            variable =  lexeme_d[(+alpha_p >> *digit_p)];
+        }
+    
+
+        rule<ScannerT> variable, select_expr, s3_object, where_clause, number, float_number, string, arith_cmp, log_op, condition_expression, arithmetic_predicate, factor;
```

Comment:
could we use a using directive or equivalently typedef here?

   using something_rule = rule<ScannerT>;

maybe?